javascript/javascript1/week3/smartPhoneApp.js

```diff
@@ -0,0 +1,141 @@
+// const formatToday = (date) => {
+//     const year = String(date.getFullYear()).slice(2);
+//     const month = String(date.getMonth() + 1);
+//     const day = String(date.getDate());
+//     return `${year}/${month}-${day}`
+// }
+// const todayFormatted = formatToday(today);
+
+// function showStatus() {
+//     if(activities.length === 0){
+//         return 'Add some activities before calling showStatus.'
+//     }else{
+//         let activitiesToday = activities.filter(activities => activities.date === todayFormatted);
+//         if(activitiesToday.length > 0){
+//             let usageOfToday = activitiesToday.reduce((totalDuration, activities) => totalDuration + activities.duration, 0);
+//             if(usageOfToday >= 120) {
+//                 return 'You have reached your limit, no more smartphoning for you!'
+//             }else{
+//                 return `You have added ${activities.length} activities. They amount to ${timeOfUsage} min. of usage.`
+//             }
+//         }else{
+//             let timeOfUsage = 0;
+//             for(i = 0; i < activities.length; i++){
+//                 timeOfUsage += activities[i].duration;
+//             }
+//             return `You have added ${activities.length} activities. They amount to ${timeOfUsage} min. of usage.`
+//         }
+//     }
+// }
+
+// console.log(showStatus());
+// // addActivity('24/11-26', 'Do Homework', 100);
+// // addActivity('24/11-26', 'Coding', 100);
+// console.log((activities));
+
+
+
+//  input Date automatically
+const activities = [];
+
+// I tried to as order of date, content,then doration, it led content and duration move forward, so here I put date in the end.
+function addActivity(activityContent, activityDurationMin, activityDateInput) { 
+    const activityDate = activityDateInput || new Date().toLocaleDateString("en-US"); // input a specify day or automatically today's date
+    let activityInfo = {
+        date: new Date(activityDate).toLocaleDateString("en-US"),
+        activity: activityContent,
+        duration: activityDurationMin,
+    };
+    activities.push(activityInfo);
+}
+addActivity("Youtube", 130, "11/26/2024");
+addActivity("Youtube", 30);
+addActivity("Facebook", 20);
+addActivity("Homework", 150)
+console.log(activities)
+
+// as date format is already unified, then showStatus should be updated. Here I would like to show status of today or a specific day.
+// here I want to alert only for the usage limit of today, if the previous records sum over limit, it won't alert but a summary of usage.
+function showStatus(dateInput) {
+    if(activities.length === 0){
+        return 'Add some activities before calling showStatus.'
+    }else{
+        let activitiesOneDay = activities.filter(activities => activities.date === dateInput);
+        if(activitiesOneDay.length > 0){
+            let usageOfOneDay = activitiesOneDay.reduce((totalDuration, activities) => totalDuration + activities.duration, 0);
```

Comment:
This is an interesting twist on the assignment! 😄 Can you briefly explain what `filter`, `reduce` and `=>` does? 🙂

It is a bit hard to follow the implementation though as there are many layers of if-statements. I would suggest looking into the "early-exiting" idiom; some quick googling should get you going! 😄 The idea is to reduce the nesting as much as possible by returning as soon as possible. A quick example is on line 104: since you returned in the then branch, you technically do not need to put the rest inside `else`. Can you see why? 🙂

Reply:
Yes, `filter` and `=>` were once mentioned in the class, when we want to find an element not knowing if there are several or just one, using 'includes' may just find one element. But here we can have several activities in one day, so  I use `filter` here.
Then the `reduce()` is used to calculate the sum of duration. When I google how to calculate the sum of numbers in object and array, most of the answers are using `reduce((accumulator, currentValue) => accumulator + currentValue, 0)`, so I tried to use here.
As you mentioned, I also feel more and more difficult to have so many conditions together. I will go to spend some time on "early-exiting". 
Because I just had a project presentation yesterday, I just start homework today 😂